controllers/ovndbcluster_controller.go

```diff
@@ -0,0 +1,389 @@
+		// update the overall status condition if service is ready
+		if instance.IsReady() {
+			instance.Status.Conditions.MarkTrue(condition.ReadyCondition, condition.ReadyMessage)
+		}
+
+		if err := helper.SetAfter(instance); err != nil {
+			util.LogErrorForObject(helper, err, "Set after and calc patch/diff", instance)
+		}
+
+		if changed := helper.GetChanges()["status"]; changed {
+			patch := client.MergeFrom(helper.GetBeforeObject())
+
+			if err := r.Status().Patch(ctx, instance, patch); err != nil && !k8s_errors.IsNotFound(err) {
+				util.LogErrorForObject(helper, err, "Update status", instance)
+			}
+		}
+	}()
+
+	// Handle service delete
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helper)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, instance, helper)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *OVNDBClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ovnv1.OVNDBCluster{}).
+		Owns(&corev1.Service{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&appsv1.StatefulSet{}).
+		Complete(r)
+}
+
+func (r *OVNDBClusterReconciler) reconcileDelete(ctx context.Context, instance *ovnv1.OVNDBCluster, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service delete")
+
+	// Service is deleted so remove the finalizer.
+	controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())
+	r.Log.Info("Reconciled Service delete successfully")
+	if err := r.Update(ctx, instance); err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *OVNDBClusterReconciler) reconcileUpdate(ctx context.Context, instance *ovnv1.OVNDBCluster, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service update")
+
+	// TODO: should have minor update tasks if required
+	// - delete dbsync hash from status to rerun it?
+
+	r.Log.Info("Reconciled Service update successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *OVNDBClusterReconciler) reconcileUpgrade(ctx context.Context, instance *ovnv1.OVNDBCluster, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service upgrade")
+
+	// TODO: should have major version upgrade tasks
+	// -delete dbsync hash from status to rerun it?
+
+	r.Log.Info("Reconciled Service upgrade successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *OVNDBClusterReconciler) reconcileNormal(ctx context.Context, instance *ovnv1.OVNDBCluster, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service")
+
+	// If the service object doesn't have our finalizer, add it.
+	controllerutil.AddFinalizer(instance, helper.GetFinalizer())
+	// Register the finalizer immediately to avoid orphaning resources on delete
+	//if err := patchHelper.Patch(ctx, openStackCluster); err != nil {
+	if err := r.Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
For safety you should return from the reconciler after doing an update. If you don't you can get read-after-write consistency errors in the reconciler (ask me how I know 😬).

The reason is that the reconciler is called with an object from a shared informer. This has a watch and a cache. As soon as you call Update() here the watch is going to update the shared informer, and this reconciler is queued to be called again *with the state of the object at this point*. i.e. Without any of the things we're adding below. We then make all the changes below and when the calling function exits we call our defer func, which updates the object again. At this point there is a race. The shared informer already has a previous reconcile queued for us. If we get called again before kube-apiserver has fired the watch from the second update and the informer has updated based on it, we will be called with old data. i.e. We write new data and then immediately get called again with a version of the object which doesn't contain the changes we just wrote. Worse, this is extremely likely to happen in practise.

Given that we're going to be called again anyway once we call update we might as well exit after a single update call and wait. It's much less likely to result in hair loss.

Reply:
Thanks Matthew and Andrew, Done

Reply:
@mdbooth you just described something very similar to my last week head scratching (documented here https://github.com/openstack-k8s-operators/nova-operator/issues/110) where I saw stale data being read at the start of Reconcile. If you know about any similar surprise factors regarding operator development please post it publicly (or point me to those posts if it is already written up somewhere).